### ISSUE TITLE
[PO] Merchants can onboard in live mode and then later onboard in dev mode, resulting in odd account issues

### ISSUE THREAD

**Issue by csmcneill** [Automattic/woocommerce-payments#6520]
### Describe the bug
If a merchant starts the onboarding process via PO but goes back to wp-admin for any reason, they will be prompted to go through the onboarding process again.

If the merchant goes through the process enough to create a live WCPay account (meaning that an `acct_` ID is assigned to that particular `blog_id`) but not enough so that the `/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fconnect` endpoint is still accessible, they can select dev mode from the available options.

When they do this, the merchant has a live account, but `wcpay_onboarding_test_mode` is set to `1`.

### To Reproduce
1. Enable PO
2. Start the onboarding process as a US account
3. When directed to the Stripe portion of the onboarding flow, return back to the WooCommerce dashboard via the browser's back button or the link in the Stripe dashboard.
4. Start the onboarding process again.
5. Select the dev mode option and proceed through the onboarding process.
6. Once onboarding is finished, navigate to the `/wp-admin/options.php` endpoint and search for `wcpay_onboarding_test_mode`.
7. Note that `wcpay_onboarding_test_mode` is set to `1`.

### Actual behavior
When `wcpay_onboarding_test_mode` is set to `1` while a live account is connected, a merchant will still be prompted to set up WCPay. This leads to pages that are inaccessible, including the inability to edit account details, until the **Payments > Overview** page is refreshed.

### Expected behavior
A few options:
1. Merchants will be directed to the portion of the onboarding process where they left off at.
2. Once a live account is connected to the `blog_id`, the initial option (live vs. dev mode) is not available.
3. `wcpay_onboarding_test_mode` is set to `0` automatically if a live account is connected to a `blog_id`.

**Comment by daquinons** [comment]
This is an awesome catch, thank you!

**Comment by brandonwanA8C** [comment]
@csmcneill is on "Hacker Mode"!

I'm trying to understand the steps to reproduce:
1) Start the onboarding process as a US account
2) When directed to the Stripe portion of the onboarding flow, return back to the WooCommerce dashboard via the browser's back button or the link in the Stripe dashboard.

Q:
a) When did the user "return" back to WC Dashboard, at the End of the Stripe flow, or in the Mid of it?
b) (the user can't see) but do we know if this user was assigned with "PO" privilege? (if the user were to finish Stripe flow with PO, they would not see the KYC/Bank acct flow before going back to WC Dash).

**Comment by csmcneill** [comment]
Hey @brandonwanA8C! You can replicate this process by returning back to the Woo dashboard immediately when the Stripe portion of onboarding begins. I believe this process may happen irrespective of whether the merchant is eligible for PO, but that would require more testing.

**Comment by dmallory42** [comment]
> I believe this process may happen irrespective of whether the merchant is eligible for PO, but that would require more testing.

Just to confirm, yes it would, but it's a little easier to replicate this bug with the new PO UX flow as we provide an option for the user to state whether they want to onboard in dev mode or live mode. (prior to this flow being introduced, the user would have to start an onboarding, go back, manually turn on dev mode and then go back to the onboarding.)

**Comment by brandonwanA8C** [comment]
knowing we (Woo & Stripe) do not have a bullet-proof integration or E2E experience, yet. this type of edge scenario will occur. Let's dive a little deeper into this scenario, so we can understand the frequency/occurrence of this behavior, to assign the proper priority.  

if my understanding is correct, from a Merchant/user's PoV:
* I'm trying to onboard a live_account
* During onboarding (while I was on the Stripe side), I changed my mind and wanted to have a test_account instead
* I abandon the Stripe flow, "reverse, backspace" back to Woo UX, to choose the "I want test_acct" radio button
**Actual:** 
* I'm seeing an "inaccessible page" and other weird behaviors

This could be a rare (non-majority) case, but I personally have encountered this: abandoning Stripe UX, coming back to Woo, and trying different onboarding settings and paths.
Since Stripe Express **does not allow the same user to be onboarded with both Live and Test accounts simultaneously**, I say let's not hack the UX to support something the backend cannot.

**Expected, given our backend capability**
IMO, until we have full flexibility to direct users to/from live/test accounts, the most preferred option:
* Once a live account is connected to the blog_id, the initial option (live vs. dev mode) is not available.

**Comment by oaratovskyi** [comment]
I explored this case, and it's interesting. A bit similar to my discovery in https://github.com/Automattic/woocommerce-payments/issues/2734#issuecomment-1718106662. As it was already mentioned: 
> Stripe Express does not allow the same user to be onboarded with both Live and Test accounts simultaneously

Except for ideas that are already on the table, I'd love to add a couple from myself:

- We could add a banner notice on the Finish Setup screen saying something like `You've started creating Stripe account, the button will redirect you where you left`
- We could just redirect our merchant to where they left off, without showing them a choice screen again. it does not make sense because no matter what the user chooses - current logic will redirect them to continue the Stripe setup where they left with the **type of account they've chosen**

IIRC second point is similar to what @brandonwanA8C suggested, correct?
> Once a live account is connected to the blog_id, the initial option (live vs. dev mode) is not available.

**Comment by oaratovskyi** [comment]
putting it on hold until we get a reply from @brandonwanA8C

**Comment by brandonwanA8C** [comment]
@oaratovskyi  
**long-term**: test & dev-mode usage & experience are a mess and a big topic; we will address & redesign the e2e experience in 2024.
**short-term**: Let's suggest a simpler solution (throw-away worthy) to ensure this behavior does not cause functional bugs or churns.

**This described behavior** sounds like a case of ambiguous usage of the Live vs. Dev modes:
_"When wcpay_onboarding_test_mode is set to 1 while a live account is connected, a merchant will still be prompted to set up WCPay."_  We need to know **why** the user would choose this and what they are **ultimately** trying to accomplish.

**My understanding of the root cause** is that: there is no "_source-of-truth_" on which type of account & UX we need to display, as there **can** be a conflict between the user-started-journey (e.g. Live path) vs. plug-in configuration (e.g. Test mode == true). 
And in the case of conflict between journey & configuration, which UX is the user committed to?

**My recommendation** for the short-term solution: we need to have separate & distinguished "Live" vs. "Test/Dev" experiences, and the experience needs to be orchestrated by the configuration (follow the Model-View-Controller concept). Therefore, IF the configuration is "TestMode == true", then the user must experience TestMode flow & UX (test onboarding if haven't started, test-KYC data, test-cards, etc). Vice versa, if "TestMode == false", then the user goes through Live flow. 

cc @elizaan36 @csmcneill

**Comment by brandonwanA8C** [comment]
One critical assumption to make and to validate is:
**Stripe Sandbox account creation is completely separate from Live account creation.**

This assumes that: just because the user has a Live acct created with Stripe, it does not automatically create an equivalent Sandbox acct without the user going through the mock (test-mode) onboarding journey in Sandbox.

@daquinons @oaratovskyi would you & team be able to confirm this via Stripe Express Live/Sandbox Documentation, or would you like to reach out to Stripe to confirm, lmk.

**Comment by oaratovskyi** [comment]
> This described behavior sounds like a case of ambiguous usage of the Live vs. Dev modes:
"When wcpay_onboarding_test_mode is set to 1 while a live account is connected, a merchant will still be prompted to set up WCPay." We need to know why the user would choose this and what they are ultimately trying to accomplish.

The most important point here is next: once one starts onboarding a Stripe account in Live mode, they get Stripe account assigned, for example, `acct_1NzPMhCLHn81eeKs`. If one clicks `Return to WooPayments` right after they see Stripe login page, they will see connect page again. They might be confused and then try Test mode, though as they already have live Stripe account assigned, they are forced to continue with this live account onboarding.

> My recommendation for the short-term solution: we need to have separate & distinguished "Live" vs. "Test/Dev" experiences, and the experience needs to be orchestrated by the configuration (follow the Model-View-Controller concept). Therefore, IF the configuration is "TestMode == true", then the user must experience TestMode flow & UX (test onboarding if haven't started, test-KYC data, test-cards, etc). Vice versa, if "TestMode == false", then the user goes through Live flow.

Gotcha! We'll start with a completely new Stripe account rather than continue previous one every time they see the screen where they need to choose live mode vs test mode. In this case merchant will have live account that we ditched onboarding and also dev account they've onboarded. We'll see if that messes things up when user decides to go live (from dev account), but that should be ok.

> One critical assumption to make and to validate is:
Stripe Sandbox account creation is completely separate from Live account creation.

Yes, it's separate.

> This assumes that: just because the user has a Live acct created with Stripe, it does not automatically create an equivalent Sandbox acct without the user going through the mock (test-mode) onboarding journey in Sandbox.

Correct!

**Comment by brandonwanA8C** [comment]
@oaratovskyi thank you for confirming the assumption, that makes the experience a lot easier to manage.

Since **Stripe Sandbox account creation is completely separate from Live account creation.**
The simplest way is to treat "Dev Mode" and "Live Mode" as two completely separate paths, like two "separate accounts" orchestrated by the plug-in configuration.
If "Dev Mode" is enabled, then the user needs to setup a sandbox account and use everything the sandbox supports (test data, test cards) until Dev Mode is disabled.

**Comment by daquinons** [comment]
Additional Slack conversation p1695745229298579-slack-C03KTTK2YMA

**Comment by brandonwanA8C** [comment]
This is my latest observation with the Dev, Test, and Live mode switching.
It seems to be a mixture of UX, Copy, and Functional issues together:

[Demo recording](https://us02web.zoom.us/rec/share/cE8uoBYwh9h9hIaSLprDIaHiX3PQZBHVCEQVbdNxRgHSBT_JfIZyHZfG6t3SKa0L.-Vlf07KQe45dW_TC?startTime=1698950243000)
Passcode: ZV@7f=!B

**Comment by daquinons** [comment]
Demo recording here p1698953366764689-slack-C03KTTK2YMA
